File: EA3/train.py
#!/usr/bin/env python
# coding: utf-8

# In[1]:


import os, json, datetime, numpy as np, tensorflow as tf
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.callbacks import TensorBoard
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, Bidirectional, LSTM, Dense


# ## Tokenize

# In[ ]:


with open("./pg1513.txt", 'r', encoding='utf-8') as file: text = file.read()
DIR = f"{os.path.splitext(os.path.basename(file.name))[0]}_{datetime.datetime.now().strftime('%Y%m%d-%H%M%S')}"
os.makedirs('./models/'+DIR, exist_ok=True)

tokenizer = Tokenizer()
tokenizer.fit_on_texts([text])
dict_size = len(tokenizer.word_index) + 1

sequences = np.array( pad_sequences( [tokenizer.texts_to_sequences([line])[0][:i] for line in text.split('\n') for i, _ in enumerate(tokenizer.texts_to_sequences([line])[0], start=1)], padding='pre') )

Y = np.array(tf.keras.utils.to_categorical(sequences[:, -1], num_classes=dict_size))

with open("./models/{0}/dict.json".format(DIR), 'w', encoding='utf-8') as f:
    json.dump({
        "word_index": tokenizer.word_index,
        "index_word": tokenizer.index_word,
        "sequence_l": sequences.shape[1]
    }, f, ensure_ascii=False
)


# ## Train

# In[ ]:


model = Sequential()
model.add(Embedding(dict_size, 100, input_shape=(sequences.shape[1]-1,)))
model.add(Bidirectional(LSTM(100)))
model.add(Dense(dict_size, activation='softmax'))
model.compile(loss='categorical_crossentropy', metrics=['accuracy'], optimizer=Adam(learning_rate=0.01))
model.summary()

history = model.fit(sequences[:, :-1], Y, epochs=1, batch_size=32, verbose=2, callbacks=[TensorBoard()])

model.save('./models/{0}/model.keras'.format(DIR))


# ## Test

# In[11]:


test = "Henceforth"

for _ in range(10): test += " " + tokenizer.index_word.get( np.argmax( model.predict( pad_sequences([(tokenizer.texts_to_sequences([test])[0])], maxlen=sequences.shape[1] - 1, padding='pre'), verbose=0) ), "")

print(test)


# ## Export

# In[12]:


import tensorflowjs as tfjs
tfjs.converters.save_keras_model(model, "./models/{0}/tfjs".format(DIR))

File: EA3/train.js
const tf = require('@tensorflow/tfjs-node-gpu');
const fs = require('fs');
const path = require('path');
const { Tokenizer } = require('tf_node_tokenizer');
const { pad_sequences } = require('./util');
const csvWriter = require('csv-writer').createObjectCsvWriter;
const { ModelCheckpoint, CsvLogger } = require('tfjs_ports')

let model;

function createDirectory(dirPath) {
    fs.mkdirSync(dirPath, { recursive: true });
}

function saveModel(model, path) {
    return model.save(`file://${path}`);
}

function loadModel(path) {
    return tf.loadLayersModel(`file://${path}/model.json`);
}

function writeToFile(path, data) {
    fs.writeFileSync(path, JSON.stringify(data));
}

/**
 * @param {number} dict_size
 * @param {number} units
 * @param {number} lr
 */
function buildModel(dict_size, units = 200, lr = 0.01) {
    let model = tf.sequential();
    model.add(tf.layers.embedding({ inputDim: dict_size, outputDim: units }));
    model.add(tf.layers.bidirectional({ layer: tf.layers.lstm({ units }) }));
    model.add(tf.layers.dense({ units: dict_size, activation: 'softmax' }));
    model.compile({ loss: 'categoricalCrossentropy', metrics: ['categoricalAccuracy'], optimizer: tf.train.adam(lr) });
    model.summary();
    return model;
}

async function trainModel() {
    // Tokenize
    const text = fs.readFileSync('./data/pg1513.txt', 'utf-8');
    const DIR = path.basename('./data/pg1513.txt').split('.')[0] + '_' + new Date().toISOString().replace(/:/g, '-');
    createDirectory('./models/' + DIR);

    const tokenizer = new Tokenizer();
    tokenizer.fitOnTexts([text]);
    const dict_size = Object.keys(tokenizer.word_index).length + 1;

    let sequences = text.split('\n').flatMap(line => {
        const sequence = tokenizer.textsToSequences([line])[0];
        return Array.from({length: sequence.length}, (_, i) => sequence.slice(0, i + 1));
    });

    const max_sequence_len = Math.max(...sequences.map(seq => seq.length));
    sequences = pad_sequences(sequences);

    const X = tf.tensor(sequences.map(seq => seq.slice(0, -1)), null, 'int32');
    const Y = tf.tensor(sequences.map(seq => seq.slice(-1)), null, 'int32');

    writeToFile(`./models/${DIR}/dict.json`, {
        word_index: tokenizer.word_index,
        index_word: tokenizer.index_word,
    });

    const writer = csvWriter({
        path: `./models/${DIR}/logs.csv`,
        header: [
            { id: 'epoch', title: 'Epoch' },
            { id: 'loss', title: 'Loss' },
            { id: 'acc', title: 'Accuracy' },
            { id: 'test', title: 'Test' }
        ],
        append: true // Append to existing file
    });
    model = buildModel(dict_size);

    let best = { acc: 0, epoch: 0 };
    const history = await model.fit(X, tf.oneHot(Y.squeeze(), dict_size), {
        epochs: process.argv.slice(2)[0] ?? 100, batchSize: 64, callbacks: {
            onEpochEnd: async function (epoch, logs) {
                let test = 'O serpent heart';
                for (let i = 0; i < 10; i++) {
                    const sequence = pad_sequences([tokenizer.textsToSequences([test])[0]], max_sequence_len - 1)[0];
                    test += " " + (tokenizer.index_word[tf.argMax(model.predict(tf.tensor([sequence])).dataSync()).dataSync()[0]] || "");
                }
                writer.writeRecords([{epoch, loss: logs.loss, acc: logs.categoricalAccuracy, test}])
                    .then(() => {
                        console.log(epoch,logs.loss,logs.categoricalAccuracy,test)
                    })
                    .catch((error) => {
                        console.error('Error writing to CSV file:', error);
                    });
                if (logs.categoricalAccuracy > best.acc) {
                    best.acc = logs.categoricalAccuracy;
                    best.epoch = epoch;
                    await saveModel(model, `./models/${DIR}/${epoch}`);
                }}, onTrainEnd: async () =>
                model = await loadModel(`./models/${DIR}/${best.epoch}`)
        }
    });
    writeToFile(`./models/${DIR}/history.json`, history);
    // You can add more lines of code here that will execute in sequence
    await saveModel(model, `./models/${DIR}/tfjs`);
    // Test
}

trainModel().catch(console.error);
